### PR TITLE
feat(thanos): adding initial plugin

### PR DESCRIPTION
**Pull request by viennaa**
---

## Pull Request Details

Plugin deploys:
* Thanos Query
* Thanos Compactor

## Breaking Changes

None

## Issues Fixed

None

## Other Relevant Information

Documentation added
Instructions on how to setup kube-monitoring provided

### CODE REVIEW

File: thanos/README.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+* Query
+<!--* Query Frontend-->
+* Compact
+<!--* (Ruler)-->
+<!--* Storegateway-->
+
+Requirements (detailed steps below):
+* ready to use credentials for a [compatible object store](https://thanos.io/tip/thanos/storage.md/)
+* thanos-sidecar enabled in Prometheus (usually with [Prometheus Operator](https://prometheus-operator.dev/))
```
Comment (richardtief) [inline thanos/README.md:17]: Pointing to the ThanosSpec directly.
```suggestion
* thanos-sidecar enabled in Prometheus (usually with [Prometheus Operator](https://prometheus-operator.dev/docs/api-reference/api/#monitoring.coreos.com/v1.ThanosSpec))
```

File: thanos/README.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+1. Tommy Sauer (@viennaa) 
+2. Richard Tief (@richardtief) 
+3. Martin Vossen (@artherd42) 
+
+# Configuration
+
+## Object Store preparation
+
+To run thanos you need object store credentials. Choose a provider and add the credentials into a secret. The [Thanos documentation](https://thanos.io/tip/thanos/storage.md/) provides a great overview on the different supported store types.
```
Comment (richardtief) [inline thanos/README.md:29]:
```suggestion
To run Thanos, you need object storage credentials. Get the credentials of your provider and add them to a Kubernetes Secret. The [Thanos documentation](https://thanos.io/tip/thanos/storage.md/) provides a great overview on the different supported store types.
```

File: thanos/README.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+**Important:** `$THANOS_PLUGIN_NAME` is needed later for the respective Thanos plugin and they must not be different!
+
+```
+kubectl create secret generic $THANOS_PLUGIN_NAME-metrics-objectstore --from-file=thanos.yaml=/path/to/your/file
+```
+
+
+## kube-monitoring plugin enablement 
+
+Prometheus in kube-monitoring needs to be altered to have a sidecar and ship metrics to the new object store too. You have to provide the secret you've just created to the (most likely already existing) kube-monitoring plugin. Add this:
```
Comment (richardtief) [inline thanos/README.md:53]:
```suggestion
Prometheus in kube-monitoring needs to be altered to have a sidecar and ship metrics to the new object store too. You have to provide the Secret you've just created to the (most likely already existing) kube-monitoring plugin. Add this:
```

File: thanos/README.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+## Thanos Querier
+
+This is the real deal now: Define your Thanos query by creating a plugin.
+
+**NOTE1:** `$THANOS_PLUGIN_NAME` needs to be consistent with your secret created earlier.
+**NOTE2:** The `releaseNamespace` needs to be the same as to where kube-monitoring resides. By default this is kube-monitoring. 
+
+
+```
```
Comment (richardtief) [inline thanos/README.md:72]:
```suggestion
```yaml
```

File: thanos/README.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  name: $YOUR_CLUSTER_NAME
+spec:
+  pluginDefinition: thanos
+  disabled: false
+  clusterName: $YOUR_CLUSTER_NAME 
+  releaseNamespace: kube-monitoring
+```
+
+## [OPTIONAL] Handling your Prometheus Store
+By default Thanos Query would check for a service `prometheus-operated` in the same namespace with this GRPC port to be available `10901`. The cli option looks like this and is configured in the PluginDefinition:
```
Comment (richardtief) [inline thanos/README.md:85]: Nice!

File: thanos/charts/values.yaml [hunk 6]
```diff
@@ -0,0 +1,92 @@
+# SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Greenhouse contributors
+# SPDX-License-Identifier: Apache-2.0
+
+global:
+
+  ## Labels to apply to all resources
+  ##
+  commonLabels: {}
+
+## kube-prometheus-stack configuration scoped to kube-monitoring
```
Comment (richardtief) [inline thanos/charts/values.yaml:10]:
```suggestion
```

File: thanos/charts/templates/compactor.yaml [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        - compact
+        - --log.level={{ default "info" .Values.thanos.compactor.logLevel }}
+        - --http-address={{ default "0.0.0.0:10902" .Values.thanos.httpAddress }}
+        - --http-grace-period={{ default "120s" .Values.thanos.compactor.httpGracePeriod }}
+        - --data-dir=/data
+        - --objstore.config-file=/etc/config/thanos.yaml
+        - --consistency-delay={{ default "1800s" .Values.thanos.compactor.consistencyDelay }}
+        - --retention.resolution-raw={{ default "7776000s" .Values.thanos.compactor.retentionResolutionRaw }}
+        - --retention.resolution-5m={{ default "7776000s" .Values.thanos.compactor.retentionResolution5m }}
+        - --retention.resolution-1h={{ default "157680000s" .Values.thanos.compactor.retentionResolution1h }}
```
Comment (richardtief) [inline thanos/charts/templates/compactor.yaml:50]: We should probably mention this in the readme, as it affects the cost of object storage quite a bit.

**Review by richardtief — CHANGES_REQUESTED** [review]
I really like that! Thanks for going to the effort of making Thanos pluggable. Some NITs below.

**Comment by richardtief** [comment]
Please rebase onto main and add the [codeowners](https://github.com/cloudoperators/greenhouse-extensions/blob/39697246d1146452fe8024032312c189248f3703/.github/CODEOWNERS) team observability to the plugin.

**Comment by IvoGoman** [comment]
The `plugin.yaml` have been renamed to `plugindefinition.yaml` across the PluginDefinitions on the main branch. Following this naming scheme is important for our CI to pick the definition up.

**Comment by viennaa** [comment]
CODEOWNERS added and plugin.yaml renamed

**Review by richardtief — APPROVED** [review]

**Review by IvoGoman — APPROVED** [review]

**Pull request merged**